### PR TITLE
#2190 Add order 1 prediction in SIM solver.

### LINKED ISSUE #2190: Order 1 prediction for SolvderSIM

### PR DESCRIPTION
**Pull request by gautierbureau**
close #2190

### CODE REVIEW

**Review by rosiereflo — CHANGES_REQUESTED** [review]
Before reviewing, please add unit tests and a testcase that illustrates this features

**Comment by rosiereflo** [comment]
Please also adapt the documentation

**Comment by gautierbureau** [comment]
Done for both comments

File: dynawo/sources/Solvers/FixedTimeStep/SolverSIM/DYNSolverSIM.cpp [hunk 1]
```diff
@@ -113,25 +151,59 @@ SolverSIM::calculateIC(double /*tEnd*/) {
 #endif
 }
 
+bool SolverSIM::hasPrediction() {
```
Comment (rosiereflo) [inline dynawo/sources/Solvers/FixedTimeStep/SolverSIM/DYNSolverSIM.cpp:154]: this method should be const

File: dynawo/sources/Solvers/FixedTimeStep/SolverSIM/DYNSolverSIM.cpp [hunk 2]
```diff
@@ -102,6 +129,17 @@ SolverSIM::calculateIC(double /*tEnd*/) {
     solverKINAlgRestoration_->solve();
     solverKINAlgRestoration_->getValues(vectorY_, vectorYp_);
 
+    if (hasPrediction()) {
+      SolverKINAlgRestoration& solverKINYPrim = getSolverKINYPrim();
+
+      // Velocity initialization with solverKINYPrim_
+      solverKINYPrim.setInitialValues(tSolve_, vectorY_, vectorYp_);
+      bool noInitSetup = false;
```
Comment (rosiereflo) [inline dynawo/sources/Solvers/FixedTimeStep/SolverSIM/DYNSolverSIM.cpp:137]: those bool cpould be const

File: dynawo/sources/Solvers/FixedTimeStep/SolverSIM/test/jobs/solverTestPrediction.dyd [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<!--
+    Copyright (c) 2015-2019, RTE (http://www.rte-france.com)
```
Comment (rosiereflo) [inline dynawo/sources/Solvers/FixedTimeStep/SolverSIM/test/jobs/solverTestPrediction.dyd:3]: 2024

File: nrt/data/SIMOrder1Prediction/Test.crv [hunk 4]
```diff
@@ -0,0 +1,19 @@
+<?xml version='1.0' encoding='UTF-8'?>
+<!--
+    Copyright (c) 2021, RTE (http://www.rte-france.com)
```
Comment (rosiereflo) [inline nrt/data/SIMOrder1Prediction/Test.crv:3]: 2024 (same for all)

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: dynawo/sources/Solvers/FixedTimeStep/DYNSolverCommonFixedTimeStep.cpp [hunk 5]
```diff
@@ -472,10 +475,30 @@ SolverCommonFixedTimeStep::reinit() {
...
+      solverKINYPrim.solve(noInitSetup, evaluateOnlyMode);
+      solverKINYPrim.getValues(vectorY_, vectorYp_);
+
+      solverKINYPrim.updateStatistics(nNewt_, nre, nje);
+      stats_.nre_ += nre;
+      stats_.nni_ += nNewt_;
+      stats_.nje_ += nje;
+    }
+
+    model_->reinitMode();
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Solvers/FixedTimeStep/DYNSolverCommonFixedTimeStep.cpp:500]: I just moved the initial one below this new piece of code

[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Solvers/FixedTimeStep/DYNSolverCommonFixedTimeStep.cpp:500]: yes, but why? Does the code above modify the modes?

File: dynawo/sources/Solvers/FixedTimeStep/SolverSIM/DYNSolverSIM.cpp [hunk 6]
```diff
@@ -83,10 +88,32 @@ SolverSIM::init(const boost::shared_ptr<Model>& model, const double t0, const do
   initCommon(model, t0, tEnd);
 }
 
+void
+SolverSIM::defineSpecificParameters() {
+  defineSpecificParametersCommon();
+
+  const bool optional = false;
+  parameters_.insert(make_pair("order1Prediction", ParameterSolver("order1Prediction", VAR_TYPE_BOOL, optional)));
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Solvers/FixedTimeStep/SolverSIM/DYNSolverSIM.cpp:96]: yes

File: dynawo/sources/Solvers/FixedTimeStep/SolverSIM/DYNSolverSIM.cpp (hunk 1 above)
Comment (gautierbureau) [reply dynawo/sources/Solvers/FixedTimeStep/SolverSIM/DYNSolverSIM.cpp:154]: done

File: dynawo/sources/Solvers/FixedTimeStep/SolverSIM/test/jobs/solverTestPrediction.dyd (hunk 3 above)
Comment (gautierbureau) [reply dynawo/sources/Solvers/FixedTimeStep/SolverSIM/test/jobs/solverTestPrediction.dyd]: done

File: nrt/data/SIMOrder1Prediction/Test.crv (hunk 4 above)
Comment (gautierbureau) [reply nrt/data/SIMOrder1Prediction/Test.crv]: done

File: dynawo/sources/Solvers/FixedTimeStep/SolverSIM/DYNSolverSIM.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/sources/Solvers/FixedTimeStep/SolverSIM/DYNSolverSIM.cpp:96]: in fact it needs to be false, the name of the parameter indicates the parameter will be optional and not its value ! Same thing is done in SolverCommonFixedTimeStep (from where I copied it !). I added a comment !

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**